pypiper/manager.py

```diff
@@ -646,244 +655,317 @@ def call_follow():
             # (or we should overwrite). create the lock (if you can)
             # Initialize lock in master lock list
-            self.locks.append(lock_file)
-            if self.overwrite_locks or recover_mode:
-                self._create_file(lock_file)
-            else:
-                try:
-                    self._create_file_racefree(lock_file)  # Create lock
-                except OSError as e:
-                    if e.errno == errno.EEXIST:  # File already exists
-                        print ("Lock file created after test! Looping again.")
-                        continue  # Go back to start
+            for lock_file in lock_files:
+                self.locks.append(lock_file)
+                if self.overwrite_locks or local_recover:
+                    self._create_file(lock_file)
+                else:
+                    try:
+                        self._create_file_racefree(lock_file)  # Create lock
+                    except OSError as e:
+                        if e.errno == errno.EEXIST:  # File already exists
+                            print ("Lock file created after test! Looping again: {}".format(
+                                lock_file))
+
+                            # Since a lock file was created by a different source, 
+                            # we need to reset this flag to re-check the locks.
+                            proceed_through_locks = False
+                            continue  # Go back to start
 
             ##### End tests block
             # If you make it past these tests, we should proceed to run the process.
 
             if target is not None:
-                print("\nTarget to produce: `" + target + "`\n")
+                print("Target to produce: {}\n".format(",".join(['`'+x+'`' for x in target])))
             else:
-                print("\nTargetless command, running...\n")
+                print("Targetless command, running...\n")
 
             if isinstance(cmd, list):  # Handle command lists
                 for cmd_i in cmd:
-                    list_ret, list_maxmem = \
-                        self.callprint(cmd_i, nofail, container)
-                    local_maxmem = max(local_maxmem, list_maxmem)
+                    list_ret, maxmem = \
+                        self.callprint(cmd_i, shell, lock_file, nofail, container)
+                    maxmem = max(maxmem) if isinstance(maxmem, Iterable) else maxmem
+                    local_maxmem = max(local_maxmem,  maxmem)
+                    list_ret = max(list_ret) if isinstance(list_ret, Iterable) else list_ret
                     process_return_code = max(process_return_code, list_ret)
 
             else:  # Single command (most common)
                 process_return_code, local_maxmem = \
-                    self.callprint(cmd, nofail, container)  # Run command
+                    self.callprint(cmd, shell, lock_file, nofail, container)  # Run command
```

Comment:
What's the guarantee of a single value for the return code? It looks like a frequent value to receive back from callprint is a list.

Reply:
https://github.com/databio/pypiper/pull/135/files#r268315433